### PR TITLE
[HGCAL] Introduce SimTracksters linking

### PR DESCRIPTION
**Pull request by lecriste**
#### PR description:

This PR:
  - introduces a vector of indexes between the two SimTracksters collections, and store it into the event;
  - improve the assignment of the regressed energy to the SimTracksters;
  - replaces the SimTracksters collection to be used in the corresponding Validation section.

No changes are expected in the output except for the "TSToCP_linking" DQM Validation section.

#### PR validation:

`runTheMatrix -l limited`

### CODE REVIEW

**Comment by AdrianoDee** [comment]
please test

**Comment by jfernan2** [comment]
@lecriste are differences like 
https://tinyurl.com/yjjearyo
https://tinyurl.com/ygvywqau
expected/understood?

I am confused from your PR description:
"No changes are expected in the output since the Validation section concerned contained already a filter on the input collection objects which screens the replacement in the third bullet above."

**Comment by lecriste** [comment]
> @lecriste are differences like
> https://tinyurl.com/yjjearyo
> https://tinyurl.com/ygvywqau
> expected/understood?

@jfernan2 yes, those differences are expected:
  - lower number of reference truth objects,
  - worse overall metrics.

In a following PR, we will fill the same Validation plots with the sibling SimTracksters collection and the metrics will improve again.


> I am confused from your PR description:
> "No changes are expected in the output since the Validation section concerned contained already a filter on the input collection objects which screens the replacement in the third bullet above."

I meant that "No changes are expected in the output **except for** the Validation section concerned", sorry for not being clear. Shall I clarify the PR description?

File: RecoHGCal/TICL/plugins/SimTrackstersProducer.cc [hunk 1]
```diff
@@ -112,23 +120,31 @@ void TrackstersFromSimClustersProducer::produce(edm::Event& evt, const edm::Even
   const auto& geom = es.getData(geom_token_);
   rhtools_.setGeometry(geom);
   auto num_simclusters = simclusters.size();
-  result->reserve(num_simclusters);
+  result->reserve(num_simclusters);  // Conservative size, will call shrink_to_fit later
+  auto num_caloparticles = caloparticles.size();
```
Comment (jpata) [inline RecoHGCal/TICL/plugins/SimTrackstersProducer.cc:124]:
```suggestion
  const auto num_caloparticles = caloparticles.size();
```

File: RecoHGCal/TICL/plugins/SimTrackstersProducer.cc [hunk 2]
```diff
@@ -95,13 +99,17 @@ void TrackstersFromSimClustersProducer::fillDescriptions(edm::ConfigurationDescr
...
 }
 
-void TrackstersFromSimClustersProducer::produce(edm::Event& evt, const edm::EventSetup& es) {
-  auto result = std::make_unique<std::vector<Trackster>>();
+void SimTrackstersProducer::produce(edm::Event& evt, const edm::EventSetup& es) {
+  auto result = std::make_unique<TracksterCollection>();
   auto output_mask = std::make_unique<std::vector<float>>();
+  auto result_fromCP = std::make_unique<TracksterCollection>();
+  auto output_mask_fromCP = std::make_unique<std::vector<float>>();
+  auto cpToSc_STS = std::make_unique<std::map<uint, std::vector<uint>>>();
```
Comment (jpata) [inline RecoHGCal/TICL/plugins/SimTrackstersProducer.cc:107]: I understand what is cpToSc (caloParticleToSimCluster), but what is STS? Perhaps being somewhat more verbose here with the variable name may help future readers of this code.

File: RecoHGCal/TICL/plugins/SimTrackstersProducer.cc [hunk 3]
```diff
@@ -112,23 +120,31 @@ void TrackstersFromSimClustersProducer::produce(edm::Event& evt, const edm::Even
...
+  result->reserve(num_simclusters);  // Conservative size, will call shrink_to_fit later
+  auto num_caloparticles = caloparticles.size();
+  result_fromCP->reserve(num_caloparticles);
 
   for (const auto& [key, lcVec] : caloParticlesToRecoColl) {
     auto const& cp = *(key);
     auto cpIndex = &cp - &caloparticles[0];
+
+    auto regr_energy = cp.energy();
+    std::vector<uint> scSTS;
```
Comment (jpata) [inline RecoHGCal/TICL/plugins/SimTrackstersProducer.cc:132]: How large can this vector grow? Can you reserve here?

**Review by jpata — COMMENTED** [review]
No reco differences observed, good. I agree that the description might be clarified to mention that the Validation particles in DQM are expected to change.
Minor code comments inline.

**Comment by lecriste** [comment]
@jpata I implemented all your comments.

**Comment by AdrianoDee** [comment]
please test

**Comment by lecriste** [comment]
> * https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35158/25106/code-checks.patch

@smuzaffar FYI this patch is incorrect.

**Comment by smuzaffar** [comment]
let me see @lecriste I am lookin in to it.

**Comment by smuzaffar** [comment]
looks like a bug in clang-tidy
```
clang-tidy -export-fixes foo.yaml -header-filter '/build/muz/tidy/CMSSW_12_1_X_2021-09-06-2300/src/.*'  src/RecoHGCal/TICL/plugins/SimTrackstersProducer.cc                                    1337 warnings generated.
RecoHGCal/TICL/plugins/SimTrackstersProducer.cc:169:13: warning: the 'empty' method should be used to check for emptiness instead of 'size' [readability-container-size-empty]
        if (result->size() < 1)
            ^~~~~~~~~~~~~~~~~~
            result->->empty()
/cvmfs/cms-ib.cern.ch/nweek-02697/slc7_amd64_gcc900/external/gcc/9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../include/c++/9.3.0/bits/stl_vector.h:1004:7: note: method 'vector'::empty() defined here
      empty() const _GLIBCXX_NOEXCEPT
```

I will open a bug report with llvm if it not already there

**Comment by jfernan2** [comment]
+1

**Comment by AdrianoDee** [comment]
+upgrade

- code changes consistent with the PR description
- following PR will address the needed changes to restore the metrics in the validation plots (e.g. [here](https://cmsweb.cern.ch/dqm/dev/session/XT4P5z))

**Comment by jpata** [comment]
+reconstruction
- for 5979802 
- technical, no relevant reco changes
- 1325.81 differences look like from https://github.com/cms-sw/cmssw/issues/32628, 11634.911 differences look like from https://github.com/cms-sw/cmssw/issues/35109

**Comment by perrotta** [comment]
+1

**Pull request merged**